### PR TITLE
Fixed string slicing

### PR DESCRIPTION
**Pull request by Kishan-Ved**
Fixes: https://github.com/lfortran/lfortran/issues/3196

### CODE REVIEW

**Comment by certik** [comment]
Can you please add a test for this too?

**Comment by Kishan-Ved** [comment]
> Can you please add a test for this too?

Done!

**Comment by Kishan-Ved** [comment]
@certik Sir is this PR good to be merged?

File: src/lfortran/semantics/ast_common_visitor.h [hunk 1]
```diff
@@ -3093,15 +3093,20 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
                 if( m_value && m_value->type == ASR::exprType::StringConstant ) {
                     ASR::StringConstant_t *m_str = ASR::down_cast<ASR::StringConstant_t>(m_value);
                     std::string sliced_str;
+                    int str_length = 0;
+                    while( m_str->m_s[str_length] != '\0' ) {
+                            str_length += 1;
+                        }
```
Comment (Shaikh-Ubaid) [inline src/lfortran/semantics/ast_common_visitor.h:3099]:
```suggestion
                        str_length += 1;
                    }
```

File: src/lfortran/semantics/ast_common_visitor.h [hunk 2]
```diff
@@ -3093,15 +3093,20 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
...
                     if( start <= 0 ) {
                         throw SemanticError("Substring `start` is less than one",
                                     loc);
                     }
+                    if(end > str_length) {
+                        throw SemanticError("Error: Substring end index at (1) exceeds the string length",
```
Comment (Shaikh-Ubaid) [inline src/lfortran/semantics/ast_common_visitor.h:3105]:
```suggestion
                        throw SemanticError("Error: Substring end index exceeds the string length",
```

Comment (Kishan-Ved) [reply src/lfortran/semantics/ast_common_visitor.h:3105]: That is the error message given by gfortran. Should we change it?

Comment (Kishan-Ved) [reply src/lfortran/semantics/ast_common_visitor.h:3105]: The error message was mentioned here: https://github.com/lfortran/lfortran/issues/3196

File: src/lfortran/semantics/ast_common_visitor.h (hunk 1 above)
Comment (Kishan-Ved) [reply src/lfortran/semantics/ast_common_visitor.h]: Done.

File: src/lfortran/semantics/ast_common_visitor.h (hunk 2 above)
Comment (Shaikh-Ubaid) [reply src/lfortran/semantics/ast_common_visitor.h:3106]: >That is the error message given by gfortran. Should we change it?

`gfortran` gives numbering for its error messages. I think `lfortran` does not number the error messages at the moment. So if you see the generated error message:

```console
semantic error: Error: Substring end index at (1) exceeds the string length
 --> tests/errors/string_slice2.f90:6:12
  |
6 |    print*, s(1: 9)
  |            ^^^^^^^ 
```
The `(1)` seems confusing as it is not labelled anywhere.

Comment (Kishan-Ved) [reply src/lfortran/semantics/ast_common_visitor.h]: Done.

**Comment by Kishan-Ved** [comment]
The check that fails gives a weird error:
```
Transaction starting
critical libmamba Download error (28) Timeout was reached [https://conda.anaconda.org/conda-forge/linux-64/nspr-4.35-h27087fc_0.conda]
    Operation too slow. Less than 30 bytes/sec transferred the last 60 seconds
Error: Process completed with exit code 1.
```

File: src/lfortran/semantics/ast_common_visitor.h [hunk 3]
```diff
@@ -3094,15 +3094,20 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
                 if( m_value && m_value->type == ASR::exprType::StringConstant ) {
                     ASR::StringConstant_t *m_str = ASR::down_cast<ASR::StringConstant_t>(m_value);
                     std::string sliced_str;
+                    int str_length = 0;
+                    while( m_str->m_s[str_length] != '\0' ) {
+                        str_length += 1;
+                    }
```
Comment (Shaikh-Ubaid) [inline src/lfortran/semantics/ast_common_visitor.h:3100]: Can we simply use `strlen()` here?
```suggestion
                    int str_length = strlen(m_str->m_s);
```

Comment (Kishan-Ved) [reply src/lfortran/semantics/ast_common_visitor.h]: Done 👍🏻

**Comment by Kishan-Ved** [comment]
There's a weird CI error again:
```
+ make -j8
Command failed.
Error: Process completed with exit code 1.
```

**Review by Shaikh-Ubaid — APPROVED** [review]
It looks good. Thanks!

File: src/lfortran/semantics/ast_common_visitor.h [hunk 4]
```diff
@@ -3094,15 +3094,17 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
                 if( m_value && m_value->type == ASR::exprType::StringConstant ) {
                     ASR::StringConstant_t *m_str = ASR::down_cast<ASR::StringConstant_t>(m_value);
                     std::string sliced_str;
+                    int str_length = strlen(m_str->m_s);
                     if( start <= 0 ) {
                         throw SemanticError("Substring `start` is less than one",
                                     loc);
                     }
+                    if(end > str_length) {
+                        throw SemanticError("Error: Substring end index exceeds the string length",
```
Comment (Shaikh-Ubaid) [inline src/lfortran/semantics/ast_common_visitor.h:3103]: Error can be removed from here
```suggestion
                        throw SemanticError("Substring end index exceeds the string length",
```

**Review by Shaikh-Ubaid — CHANGES_REQUESTED** [review]
I just shared a minor concern. I think this would be good to merge once that is addressed.

**Comment by Kishan-Ved** [comment]
> I just shared a minor concern. I think this would be good to merge once that is addressed.

Done.

File: src/lfortran/semantics/ast_common_visitor.h (hunk 4 above)
Comment (Kishan-Ved) [reply src/lfortran/semantics/ast_common_visitor.h]: Updated.

**Review by Shaikh-Ubaid — APPROVED** [review]
Looks good. Thanks!

**Pull request merged**